[TIP] Enables TI plugin with kibana.yml feature flag

PhilippeOberti opened the pull request:
## Summary

The Threat Intelligence plugin (introduced in [this PR](https://github.com/elastic/kibana/pull/136479)) and is not yet ready for General Availability. This PR hides the new plugin behind a feature flag. As the plugin is dynamically loaded within the Security Solution plugin, we're leveraging the already existing `allowedExperimentalValues` constant to:
- show/hide the navbar entries in both old and new navigation
- load the plugin

### How to test

Threat Intelligence plugin disabled:

- start Kibana without adding any Threat Intelligence entry to your [kibana.yml](https://github.com/elastic/kibana/blob/main/config/kibana.yml)
- check that the in the old navigation, the Threat Intelligence entry doesn't appear in the Explore section of the sidenav
- check that in the new navigation, the Threat Intelligence panel doesn't appear in the Explore section
- check that trying to deep link to http://localhost:5601/app/security/threat_intelligence redirects to http://localhost:5601/app/security/get_started

https://user-images.githubusercontent.com/17276605/182380842-f12e2a49-4128-4531-9218-ae747fd3613a.mov

Threat Intelligence plugin enabled:

- in your [kibana.yml](https://github.com/elastic/kibana/blob/main/config/kibana.yml) file, add the following entry at the bottom `xpack.securitySolution.enableExperimental: ['threatIntelligenceEnabled']`
- check that 

https://user-images.githubusercontent.com/17276605/182380964-72016ab2-3c44-4eec-9c59-2d2d853ca48b.mov

When the Threat Intelligence plugin is ready for General Availability, we should be able to simply do a revert of this commit! :) 

https://github.com/elastic/security-team/issues/4505

### Checklist

Delete any items that are not applicable to this PR.

- [ ] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

<!--ONMERGE {"backportTargets":["8.4"]} ONMERGE-->

PhilippeOberti committed:
[TIP] Enables TI plugin with kibana.yml feature flag

- show/hide navbar entries in both old and new navigation per feature flag value
- load plugin per feature flag value
- flag: xpack.securitySolution.enableExperimental: ['threatIntelligenceEnabled']

https://github.com/elastic/security-team/issues/4505

PhilippeOberti commented on x-pack/plugins/security_solution/common/experimental_features.ts:
```diff
@@ -22,9 +22,10 @@ export const allowedExperimentalValues = Object.freeze({
   pendingActionResponsesWithAck: true,
   policyListEnabled: true,
   policyResponseInFleetEnabled: true,
+  threatIntelligenceEnabled: false,
```
not sure I was allowed to use this object but it seemed to make sense, as the Threat Intelligence plugin is loaded within the Security Solution plugin

PhilippeOberti commented on x-pack/plugins/security_solution/public/common/links/app_links.ts:
```diff
@@ -30,17 +32,27 @@ export const links = Object.freeze([
 ): Promise<AppLinkItems> => {
   const managementFilteredLinks = await getManagementFilteredLinks(core, plugins);
 
+  const threatHuntingFilteredLinks = {
```
I hesitated to go the opposite way (adding the threatIntelligence entry to the threatHuntingLandingLinks.link property) but going this current route meant modifying one less file.

semd replied:
no need to do this, links have support for experimentalFeatures, you can add the `experimentalKey: 'threatIntelligenceEnabled',` prop directly to the threat intelligence link definition at: https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/public/threat_intelligence/links.ts

PhilippeOberti replied:
hahaha I had took a not of this `experimentalKey` while working on another feature, tried it and didn't get it to work... I'll try again thank you :)

PhilippeOberti replied:
yeah that worked, I don't know what I did... must have forgotten to update my kibana.yml... 🤦‍♂️

lgmys commented:
How does it work with our e2e suite?:)

lgmys commented on x-pack/plugins/security_solution/public/types.ts:
```diff
@@ -148,5 +148,5 @@ export interface StartedSubPlugins {
   management: ReturnType<Management['start']>;
   landingPages: ReturnType<LandingPages['start']>;
   cloudSecurityPosture: ReturnType<CloudSecurityPosture['start']>;
-  threatIntelligence: ReturnType<ThreatIntelligence['start']>;
```
lets add a word or two why this is optional :)

PhilippeOberti replied:
good idea, I can add a comment that this is a temporary change!

PhilippeOberti replied:
added!

PhilippeOberti commented:
> How does it work with our e2e suite?:)

crap I forgot about that, looking into it now 😆

PhilippeOberti committed:
[TIP] Add config entry for e2e Cypress tests

PhilippeOberti commented:
> How does it work with our e2e suite?:)

fixed with this https://github.com/elastic/kibana/pull/137838/files#diff-44620cc130a0b96260224571a9a76e86eae76d5924f3798cca8838224bf8340cR54

maxcold reviewed (approved):
Tested, works as expected

PhilippeOberti commented:
I just found this `hideWhenExperimentalKey` [here](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/public/app/deep_links/index.ts#L114), need to investigate if I could use this and simplify some logic

PhilippeOberti committed:
[TIP] Use experimentalKey feature instead of filtering out entry manually

semd commented:
Would it be possible to start the subPlugin normally in the plugin and check the experimental flag inside the threat_intelligence main component?
https://github.com/elastic/kibana/blob/70efbf0ea3ec6eecd4663d53912ab01b7184c117/x-pack/plugins/security_solution/public/threat_intelligence/routes.tsx#L14-L25

So we don't have to do all those conditionals in the `plugin.tsx` and the types file. It will be easier to undo later on as well.
Suggestion:

```
import { useIsExperimentalFeatureEnabled } from '../common/hooks/use_experimental_features';

const ThreatIntelligenceRoutes = () => {
  const enabled = useIsExperimentalFeatureEnabled('threatIntelligenceEnabled');
  if (!enabled) {
    return null;
  }
  return (
    <TrackApplicationView viewId={SecurityPageName.threatIntelligence}>
      <ThreatIntelligencePage />
    </TrackApplicationView>
  );
};

export const routes: SecuritySubPluginRoutes = [
  {
    path: THREAT_INTELLIGENCE_PATH,
    component: ThreatIntelligenceRoutes,
  },
];
```

PhilippeOberti commented:
> Would it be possible to start the subPlugin normally in the plugin and check the experimental flag inside the threat_intelligence main component?

@semd absolutely! I thought I was being smart by not having the plugin load at all, but if this is not required/desired I'll gladly remove all the logic and make this PR a lot smaller!

PhilippeOberti committed:
[TIP] Simplified TI plugin logic

Plugin is now loaded all the time, and the logic to show the indicators page or route back to the Get Started page from Security is handled within the Threat Intelligence folder

semd reviewed (approved):
Awesome! LGTM 🚀 
thanks

MadameSheema commented on x-pack/test/security_solution_cypress/config.ts:
```diff
@@ -52,6 +52,7 @@ export default async function ({ readConfigFile }: FtrConfigProviderContext) {
           'riskyHostsEnabled',
           'riskyUsersEnabled',
           'insightsRelatedAlertsByProcessAncestry',
+          'threatIntelligenceEnabled',
```
Is there any plan of adding specific TI plugin tests on Security Solution Cypress tests?

PhilippeOberti replied:
@MadameSheema we have some e2e Cypress tests [in our plugin directly](https://github.com/elastic/kibana/blob/main/x-pack/plugins/threat_intelligence/cypress/integration/sources/indicators.spec.ts) and [this PR](https://github.com/elastic/kibana/pull/137954) updates [some e2e Cypress tests](https://github.com/elastic/kibana/pull/137954/files#diff-2cb85ba5d1effa234d5f55e4d4bfe1853ff6f97478ab0f36fbe95c9bb3f6c6b1R147) within the Security Solution plugin.

I wasn't thinking about adding any other tests, especially because this feature flag is temporary and _should_ be very short lived. We're planning on releasing everything in 8.5 (plans could change though).

Did you have anything specify in mind you wanted me/us to add?

PhilippeOberti commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into tip-feature-flag

MadameSheema reviewed (approved):
LGTM

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.4|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/138373">](https://github.com/elastic/kibana/pull/138373)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)

<!--- Backport version: 8.9.0 -->
